File: jediterm-core-mpp/src/jvmMain/java/com/jediterm/terminal/ProcessTtyConnector.java
package com.jediterm.terminal;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.Collections;
import java.util.List;

/**
 * @author traff
 */
public abstract class ProcessTtyConnector implements TtyConnector {
  protected final InputStream myInputStream;
  protected final OutputStream myOutputStream;
  protected final InputStreamReader myReader;
  protected final Charset myCharset;
  private final Process myProcess;
  private final @Nullable List<String> myCommandLine;

  public ProcessTtyConnector(@NotNull Process process, @NotNull Charset charset) {
    this(process, charset, null);
  }

  public ProcessTtyConnector(@NotNull Process process, @NotNull Charset charset, @Nullable List<String> commandLine) {
    myOutputStream = process.getOutputStream();
    myCharset = charset;
    myInputStream = process.getInputStream();
    myReader = new InputStreamReader(myInputStream, charset);
    myProcess = process;
    myCommandLine = commandLine;
  }

  @NotNull
  public Process getProcess() {
    return myProcess;
  }

  @Override
  public abstract String getName();

  public @Nullable List<String> getCommandLine() {
    return myCommandLine != null ? Collections.unmodifiableList(myCommandLine) : null;
  }

  public int read(char[] buf, int offset, int length) throws IOException {
    return myReader.read(buf, offset, length);
  }

  public void write(byte[] bytes) throws IOException {
    myOutputStream.write(bytes);
    myOutputStream.flush();
  }

  @Override
  public boolean isConnected() {
    return myProcess.isAlive();
  }

  @Override
  public void write(String string) throws IOException {
    write(string.getBytes(myCharset));
  }

  @Override
  public void close() {
    myProcess.destroy();
    try {
      myOutputStream.close();
    }
    catch (IOException ignored) { }
    try {
      myInputStream.close();
    }
    catch (IOException ignored) { }
  }

  @Override
  public int waitFor() throws InterruptedException {
    return myProcess.waitFor();
  }

  @Override
  public boolean ready() throws IOException {
    return myReader.ready();
  }
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/ProcessTtyConnector.kt
package com.jediterm.terminal

import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.nio.charset.Charset
import java.util.*

/**
 * @author traff
 */
abstract class ProcessTtyConnector @JvmOverloads constructor(
    val process: Process,
    protected val myCharset: Charset,
    private val myCommandLine: MutableList<String?>? = null
) : TtyConnector {
    protected val myInputStream: InputStream
    protected val myOutputStream: OutputStream
    protected val myReader: InputStreamReader

    init {
        myOutputStream = process.getOutputStream()
        myInputStream = process.getInputStream()
        myReader = InputStreamReader(myInputStream, myCharset)
    }

    abstract override val name: String?

    val commandLine: MutableList<String?>?
        get() = if (myCommandLine != null) Collections.unmodifiableList<String?>(myCommandLine) else null

    @Throws(IOException::class)
    override fun read(buf: CharArray?, offset: Int, length: Int): Int {
        return myReader.read(buf, offset, length)
    }

    @Throws(IOException::class)
    override fun write(bytes: ByteArray?) {
        bytes?.let {
            myOutputStream.write(it)
            myOutputStream.flush()
        }
    }

    override val isConnected: Boolean
        get() = process.isAlive()

    @Throws(IOException::class)
    override fun write(string: String?) {
        string?.let { write(it.toByteArray(myCharset)) }
    }

    override fun close() {
        process.destroy()
        try {
            myOutputStream.close()
        } catch (ignored: IOException) {
        }
        try {
            myInputStream.close()
        } catch (ignored: IOException) {
        }
    }

    @Throws(InterruptedException::class)
    override fun waitFor(): Int {
        return process.waitFor()
    }

    @Throws(IOException::class)
    override fun ready(): Boolean {
        return myReader.ready()
    }
}
